File: 11_Container With Most Water.java
// Time complexity: O(n)
// Space complexity: O(1)

class Solution {
    public int maxArea(int[] height) {
        int left = 0, right = height.length - 1, area = 0;

        // shift the pointer with lower boundary
        while (left < right) {
            area = Math.max(area, Math.min(height[left], height[right]) * (right - left));
            // lower left wall
            if (height[left] < height[right]) left ++;
            // lower right wall
            else right --;
        }

        return area;
    }
}

File: 11_Container With Most Water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        # width: r-l; height: min(h[l], h[r]) -> shift the lower boundary
        res, l, r = 0, 0, len(height) - 1

        while l < r:
            res = max(res, (r-l)*(min(height[l], height[r])))
            if height[l] < height[r]: l += 1
            else: r -= 1
        return res